fault_proof/bin/challenger.rs

```diff
@@ -0,0 +1,258 @@
+    T: Transport + Clone,
+{
+    config: ChallengerConfig,
+    l1_provider: L1Provider,
+    l2_provider: L2Provider,
+    l1_provider_with_wallet: FillProvider<F, P, T, Ethereum>,
+    factory: DisputeGameFactoryInstance<T, L1ProviderWithWallet<F, P, T>>,
+    proof_reward: U256,
+}
+
+impl<F, P, T> OPSuccicntChallenger<F, P, T>
+where
+    F: TxFiller<Ethereum>,
+    P: Provider<T, Ethereum> + Clone,
+    T: Transport + Clone,
+{
+    pub async fn new(
+        l1_provider_with_wallet: FillProvider<F, P, T, Ethereum>,
+        factory: DisputeGameFactoryInstance<T, L1ProviderWithWallet<F, P, T>>,
+    ) -> Result<Self> {
+        let config = ChallengerConfig::from_env()?;
+        let l1_provider = ProviderBuilder::default().on_http(config.l1_rpc.clone());
+
+        Ok(Self {
+            config: config.clone(),
+            l1_provider: l1_provider.clone(),
+            l2_provider: ProviderBuilder::default().on_http(config.l2_rpc.clone()),
+            l1_provider_with_wallet: l1_provider_with_wallet.clone(),
+            factory: factory.clone(),
+            proof_reward: factory
+                .fetch_proof_reward(config.game_type, l1_provider)
+                .await?,
+        })
+    }
+
+    async fn challenge_game(&self, game_address: Address) -> Result<()> {
+        let game =
+            OPSuccinctFaultDisputeGame::new(game_address, self.l1_provider_with_wallet.clone());
+
+        let receipt = game
+            .challenge()
+            .value(self.proof_reward)
+            .send()
+            .await?
+            .get_receipt()
+            .await?;
+
+        tracing::info!(
+            "Successfully challenged game {:?} with tx {:?}",
+            game_address,
+            receipt.transaction_hash
+        );
+
+        Ok(())
+    }
+
+    pub async fn get_oldest_challengable_game_address(&self) -> Result<Option<Address>> {
+        // Get latest game index, return None if no games exist
+        let Some(latest_game_index) = self.factory.fetch_latest_game_index().await? else {
+            tracing::info!("No games exist yet");
+            return Ok(None);
+        };
+
+        // Start from the latest game index - max_games_to_check_for_challenge
+        let mut game_index = latest_game_index
+            .saturating_sub(U256::from(self.config.max_games_to_check_for_challenge));
+        let mut game_address;
+        let mut block_number;
+
+        loop {
+            // If we've reached last index and still haven't found a valid proposal
+            if game_index > latest_game_index {
+                tracing::info!("No invalid proposals found after checking all games");
+                return Ok(None);
+            }
+
+            game_address = self.factory.fetch_game_address_by_index(game_index).await?;
+            let game = OPSuccinctFaultDisputeGame::new(game_address, self.l1_provider.clone());
+
+            let claim_data = game.claimData().call().await?.claimData_;
```

Comment:
Why does this have an _ at the end? Is it just the alloy type that's returned?

Reply:
It's all from the contract.rs inside the sol! macro
```
    #[allow(missing_docs)]
    #[sol(rpc)]
    contract OPSuccinctFaultDisputeGame {
        function l2BlockNumber() public pure returns (uint256 l2BlockNumber_);
        function rootClaim() public pure returns (Claim rootClaim_);
        function status() public view returns (GameStatus status_);
        function claimData() public view returns (ClaimData memory claimData_);
        function challenge() external payable returns (ProposalStatus);
        function resolve() external returns (GameStatus status_);
        function genesisL2BlockNumber() external view returns (uint256 genesisL2BlockNumber_);
        function proofReward() external view returns (uint256 proofReward_);
    }
```